[DropIn] InfoPanel, BackPress, MapMarkers and UseCases

tomaszrybakiewicz opened the pull request:
> This is a big PR that still has few TODOs left. I am creating it now to get early feedback. 

Closes [#1438](https://github.com/mapbox/navigation-sdks/issues/1438)

### Description

Besides InfoPanel components, this PR also introduces new coordinators, utilities, factories and use case classes for business logic processing.

New Coordinators:
- `BackPressedCoordinator` that manages onBackPressedCallback enabled state and handles onBackPressed event for each NavigationState.
- `NavigationStateCoordinator` coordinator that observes `viewModel.destination`,`viewModel.activeNavigationStarted` and `mapboxNavigation.flowRoutesUpdated()` flows and updates NavigationState as following:

  ```
   [ FreeDrive ]━━( destination != null )━━━━━━>>[ RoutePreview ]<<━━━━━━━━━━━━┓
        ^         ( & routes not empty  )               ┃                      ┃
        ┃                                  (activeNavigationStarted=true)      ┃
        ┃                                               ┃                      ┃
        ┃                                               ┃         (activeNavigationStarted=false)
        ┣━━━━( destination == null )━━━━┓               v                      ┃
        ┗━━━━( routes empty        )━━━━╋━━━━━[ ActiveNavigation ]━━━━━━━━━━━━━┫
                                        ┃               ┃                      ┃
                                        ┃  (onFinalDestinationArrival())       ┃
                                        ┃               ┃                      ┃
                                        ┃               v                      ┃
                                        ┗━━━━━━━━━━[ Arrival ]━━━━━━━━━━━━━━━━━┛
  ```

- `RouteCoordinator` that fetches and sets new route on `viewModel.destination` updates when in RoutePreview state.

New Components:
- `InfoPanelFreeDriveComponent` that controls InfoPanel when in FreeDrive state.
- `InfoPanelRoutePreviewComponent` that controls InfoPanel when in RoutePreview state.
- `InfoPanelActiveGuidanceComponent` that controls InfoPanel when in ActiveGuidance state .
- `MapMarkersComponent` for rendering all drop-in UI map markers.
- `LongPressMapComponent` for updating `viewModel.destination` on map long press gesture.

New Utilities, Factories, Extensions
- `MapMarkerFactory` factory class for creating all drop-in UI map point annotations.
- `BitmapMemoryCache` [LruCache](https://developer.android.com/reference/android/util/LruCache) based Bitmap memory cache. 
- Added various factory methods to `DropInNavigationViewContext`
- Added coroutine extension methods for `LocationEngine` and `MapboxNavigation`

New assets:
- route preview vector icon
- start navigation vector icon
- time/eta vector icon
- exit/close vector icon
- new styles and drop-in color resources

New Use Cases:
- `UseCase` base use case class that executes business logic synchronously or asynchronously using Coroutines. This class has been cloned  from [google/iosched project](https://github.com/google/iosched/blob/main/shared/src/main/java/com/google/samples/apps/iosched/shared/domain/CoroutineUseCase.kt) thats under Apache License, Version 2.0. 
- `FlowUseCase` base use case class that executes business logic in its execute method and keep posting updates via returned Flow.  This class has been cloned from [google/iosched project](https://github.com/google/iosched/blob/main/shared/src/main/java/com/google/samples/apps/iosched/shared/domain/FlowUseCase.kt) thats under Apache License, Version 2.0. 
- `GetCurrentLocationUseCase` use case for retrieving latest device location.
- `FetchRouteUseCase` use case for requesting routes between current location and destination point.
- `FetchAndSetRouteUseCase` use case for requesting and setting new routes to `MapboxNavigation`
- `StartActiveGuidanceUseCase` use case that starts `MapboxNavigation.mapboxReplayer` and updates shared `activeNavigationStarted` state 
- `StopActiveGuidanceUseCase` use case that stops `MapboxNavigation.mapboxReplayer` and updates shared `activeNavigationStarted` state 

Other changes:
- Added unbind() method to Binder interface.
- Updated `UICoordinator` to call both `bind` and `unbind`.

### Screencaps

https://user-images.githubusercontent.com/2678039/155572150-a2024ea4-8a49-4585-812c-fca17f329e1f.mp4

**Known issues**
- Route line is not cleared on `MapboxNavigation.setRoutes(emptyList())`.
- Camera behaviour when transitioning between NavigationStates.

### TODOs
- [x] unit tests
- [x] arrival info panel
- [x] integrate with TripProgressView
- [x] ~~implement POI Name update logic~~ (will be done in a separate PR)
- [x] add view injection support
- [x] finalize `OnBackPressedCallback` installation (currently it's done manually in host activity).
- [x] code cleanup

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt:
```diff
@@ -136,56 +137,21 @@ class DropInNavigationView @JvmOverloads constructor(
             MapCoordinator(navigationContext, binding.mapView),
             GuidanceCoordinator(navigationContext, binding.guidanceLayout),
-            InfoPanelCoordinator(navigationContext, binding.infoPanelLayout),
+            InfoPanelCoordinator(navigationContext, binding.infoPanelLayout, binding.guideBottom),
```
Why do we need a `buideBottom`? The `infoPanelLayout` was supposed to encompass "anything at the bottom" for our portrait layout.

tomaszrybakiewicz replied:
My understanding is that `infoPanelLayout` is only BottomSheet.

So I moved `roadNameLayout` to  the `ConstraintLayout` and  anchored it to horizontal `Guideline` which position is updated by `BottomSheetBehavior.BottomSheetCallback` in InfoPanelCoordinator. I think this solution will not only allow us to position `roadNameLayout` above bottom sheet but will also allow us to freeze `roadNameLayout` position when bottom sheet is dragged to a fullscreen.

kmadsen replied:
ah was confused `guideBottom`. maybe `infoPanelGuideline`

tomaszrybakiewicz replied:
It's not confusing to me. I will rename it to `guidelineBottom`.

tomaszrybakiewicz commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewContext.kt:
```diff
@@ -25,4 +42,60 @@ internal class DropInNavigationViewContext(
+        navigation,
+        viewModel._activeNavigationStarted,
+        Dispatchers.Main
+    )
```
This is a ton of boilerplate code that can be simplified with Dagger 2. What do you guys think?

tomaszrybakiewicz replied:
See https://github.com/mapbox/mapbox-navigation-android/pull/5517

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/coordinator/NavigationStateCoordinator.kt:
```diff
@@ -0,0 +1,48 @@
+ */
+internal class NavigationStateCoordinator(
+    context: DropInNavigationViewContext
+) : UIComponent() {
```
I'm only commenting on the naming here. `UICoordinator` is a class, and the definition of the "coordinator" is tied closer to that class than other possible philosophies we try to fit it to. 

We are calling other components a "behavior", that extend `UIComponent`. We should call this `NavigationStateBehavior`.

tomaszrybakiewicz replied:
To help me better understand differences between `Coordinator`, `Component` and `Behaviour` classes, can you give me a brief description for each type?

To be honest with you, "behaviour" also doesn't fit this class purpose. Behaviour classes are _Adjustment_ type of classes that when assigned to a target, changes or augments its basic functionality. This class observes 3 sources of truth (destination, routes and activeNavigationStarted) and moves NavigationState to a correct state by following above FSM. I think either "Controller" or "Manager" would be more appropriate name here.

kmadsen replied:
Linking some definitions https://github.com/mapbox/mapbox-navigation-android/pull/5521 in case someone comes in for history

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/usecase/route/FetchRouteUseCase.kt:
```diff
@@ -0,0 +1,31 @@
+import kotlinx.coroutines.CoroutineDispatcher
+
+class FetchRouteUseCase(
+    private val navigation: MapboxNavigation,
```
Why not make these `MapboxNavigationObserver`?

The `MapboxNavigation` instance can change at any time. So these use cases need to be re-instantiated?

tomaszrybakiewicz replied:
These classes encapsulate business logic that drive the system. They should be lightweight, stateless and only execute their logic when invoked.

`MapboxNavigation` is this class direct dependency. When it changes, this class must be reinstantiated. If we find that instantiation of this class affects system performance we will address it by injecting `Provider<MapboxNavigation>` class that will return currently attached `MapboxNavigation` during execution.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewContext.kt:
```diff
@@ -25,4 +42,60 @@ internal class DropInNavigationViewContext(
+        override fun handleOnBackPressed() = viewModel.onBackPressed()
+    }
+
+    val navigation: MapboxNavigation get() = checkNotNull(MapboxNavigationApp.current()) {
+        "MapboxNavigationApp not initialized. You must call MapboxNavigationApp.setup() first."
+    }
```
I think we should delete this

tomaszrybakiewicz replied:
Why? This is just an alias with a precondition to ensure that `MapboxNavigationApp.setup()` has been called.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/coordinator/RouteCoordinator.kt:
```diff
@@ -0,0 +1,32 @@
+) : UIComponent() {
+
+    private val viewModel = context.viewModel
+    private val fetchAndSetRouteUseCase: FetchAndSetRouteUseCase by lazy {
+        context.fetchAndSetRouteUseCase()
+    }
```
This will hold onto old instances of `MapboxNavigation`.

tomaszrybakiewicz commented on examples/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationViewActivity.kt:
```diff
@@ -11,6 +11,8 @@ class MapboxNavigationViewActivity : AppCompatActivity() {
         val binding = LayoutActivityNavigationViewBinding.inflate(layoutInflater)
         setContentView(binding.root)
 
+        onBackPressedDispatcher.addCallback(this, binding.root.getOnBackPressedCallback())
```
Few ideas to automate this.
1. Add `MapboxNavigation` extension method that will add `OnBackPressedCallback` to `AppCompatActivity.onBackPressedDispatcher`.
```kotlin
fun MapboxNavigation.initDropInUI(activity: AppCompatActivity, navView: DropInNavigationView) {
    activity.onBackPressedDispatcher.addCallback(activity, navView.getOnBackPressedCallback())
}
```
2. Add `AppCompatActivity` extension method that takes only `DropInNavigationView` arg.
```kotlin
fun AppCompatActivity.initDropInUI(navView: DropInNavigationView) {
    onBackPressedDispatcher.addCallback(this, navView.getOnBackPressedCallback())
}
```
3. Use Application.ActivityLifecycleCallbacks and AndroidManifest to `<meta-data>` tags to identify which activity is a NavigationView host and for which we should  register OnBackPressedCallback to intercept onBackPress. This solution would only work with activities that subclass `AppCompatActivity`.

kmadsen replied:
We can also capture the back stack inside the view. This could remove the need to add anything to the hosting activity/fragment

Prototyped an idea and [left it in documentation](https://docs.google.com/document/d/1u7T_tuTBK2iLHr6yhYQOSjb7ldVDy6rDuiz0kQCMn7I/edit?usp=sharing)
``` kotlin
@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
class NavigationBackStack(val view: View) : MapboxNavigationObserver {
    override fun onAttached(mapboxNavigation: MapboxNavigation) {
        view.isFocusableInTouchMode = true
        view.requestFocus()
        view.setOnKeyListener { _, keyCode, event ->
            if (event.action == KeyEvent.ACTION_UP && keyCode == KeyEvent.KEYCODE_BACK) {
                logger.e(TAG, Message("TODO handle back pressed"))
                false
            } else {
                false
            }
        }
    }

    override fun onDetached(mapboxNavigation: MapboxNavigation) {
        view.isFocusableInTouchMode = false
        view.setOnKeyListener { _, _, _ -> false }
    }

    private companion object {
        private val TAG = Tag("MbxNavigationBackStack")
    }
}
```

tomaszrybakiewicz replied:
Let do this in a separate PR.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/usecase/UseCase.kt:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Executes business logic synchronously or asynchronously using Coroutines.
+ */
+abstract class UseCase<in P, R>(private val coroutineDispatcher: CoroutineDispatcher) {
```
It may be good to introduce a "single shot". But we wouldn't want a public api for exposing a generic use case like this inside the `MapboxNavigation` domain. The iosched samples repository is a domain specific SDK, so it makes sense there.

What do you think about adding this. It could work along with `MapboxNavigationObserver`
``` kotlin
object MapboxNavigationApp {
...
  fun invoke(single: MapboxNavigationSingle, coroutineDispatcher: CoroutineDispatcher)
```

``` kotlin
interface MapboxNavigationSingle {
  suspend fun invoke(mapboxNavigation: MapboxNavigation)
}
```

tomaszrybakiewicz replied:
All UseCases (including abstract base classes) will be internal and won't be exposed to the public.
I am not worried about the visibility just yet. We will address it later when we are closer to the release.

`MapboxNavigationSingle` is an interesting approach however I don't see a reason why we would need this. When `MapboxNavigation` changes, we can just re-instantiate UseCase instance. 
Take a look at `Provider<StopActiveGuidanceUseCase>` injection here: https://github.com/mapbox/mapbox-navigation-android/pull/5517/files#diff-b527260b6e65ce49735cfb9ee33bd4951a61723b211330761ebdc71db984c9d7
Notice that `stopActiveGuidanceUseCaseProvider.get().invoke(Unit)` will reinstantiate StopActiveGuidanceUseCase with currently attached MapboxNavigation instance before invoking it.

kmadsen replied:
Will you make the file `internal` then?

The `MapboxNavigationObserver` is public, and eventually some of the dropin ui "components" will be public. At least that was my intention. That way, developers can re-use some of the things we build.

If we intend to always keep `UseCase` internal, then is the intention to keep the philosophy for them we keep very specific to `NavigationView`?

tomaszrybakiewicz replied:
Yes. I will make it internal.
It's to early to say if we ever make them public. I will leave it up to leads to decide.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/extensions/coroutines/MapboxNavigation.kt:
```diff
@@ -0,0 +1,48 @@
+package com.mapbox.navigation.dropin.extensions.coroutines
```
nit: Rename `MapboxNavigation.kt` to something unique

`MapboxCoroutineExtensions.kt`, and delete the `coroutines` package?

tomaszrybakiewicz replied:
What's wrong with packages? Why unique filename? We never import extension methods by filename.

tomaszrybakiewicz replied:
Better IDE integration? What IDE are you using? 
In Android Studio you can select "Classes" tab in the search dialog.

![Screen Shot 2022-03-02 at 3 24 56 PM](https://user-images.githubusercontent.com/2678039/156443748-f5556a7c-3247-40ee-8031-edf6ed0ad13b.png)

As for the Files... 

![Screen Shot 2022-03-02 at 3 23 18 PM](https://user-images.githubusercontent.com/2678039/156443586-fcf16290-cea6-48ed-9a95-4459afd58534.png)

...yeah... you are right... this is super confusing. How will we ever figure out which file is which?
<img width="250" src="https://c.tenor.com/o5CbgYywhaoAAAAd/justin-timberlake-stare.gif" />

LukasPaczos replied:
We are only one extension function or data class away from the `core.MapboxNavigation` becoming a file and not a class which would not allow us to filter it out this way anymore.

What's the problem with adding a prefix/suffix with the desired purpose of the new file? I don't understand why we should go through this cognitive hoops and compare packages when we search.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/extensions/coroutines/MapboxNavigation.kt:
```diff
@@ -0,0 +1,48 @@
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+data class RoutesWithOrigin(
+    val routes: List<DirectionsRoute>,
+    val routerOrigin: RouterOrigin
+)
+
+class RequestRoutesError(
+    val reasons: List<RouterFailure>,
+    message: String
+) : Error(message)
```
We would not make these public at any point right? 

Make them `internal`. Also thinking they should probably not exist 🤔 .

tomaszrybakiewicz replied:
Correct. They will be internal. I am not worried about the visibility just yet. We will address it later when we are closer to the release.

We need better, coroutine based interface for route fetching. The MapboxNavigation only provides archaic, callback based API that doesn't fit our needs.

`MapboxNavigation.requestRoutes()` returns multiple values via `onRoutesReady(routes: List<DirectionsRoute>, routerOrigin: RouterOrigin)` callback. The suspendible coroutine must return both success values as a result. So, to do so we can either return both values via kotlin [Pair](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-pair/)\<List\<DirectionsRoute\>, RouterOrigin\> data class or custom value object that better fits our domain. 
What approach would you choose?

cafesilencio replied:
`MapboxDropInRouteRequester` and `RouteFetchComponent` have already been merged in to handle route requesting/fetching for the drop-in UI.

tomaszrybakiewicz replied:
@cafesilencio I know it is. This PR proposes alternative approach.

LukasPaczos replied:
> I am not worried about the visibility just yet. We will address it later when we are closer to the release.

That's a valid way to think about it, but please default to least possible visibility, we can always go through the code and expose the things that should be public. It'd be very likely for us to miss some helpers, data classes, or other obscure structures and release them publicly unintentionally if we default to public and try to hide later.

abhishek1508 commented:
Thank you for drawing the state diagram @tomaszrybakiewicz :slightly_smiling_face:
However, it's slightly incorrect. This is the correct one:
<img width="400" alt="Screen Shot 2022-02-25 at 2 58 38 PM" src="https://user-images.githubusercontent.com/9770186/155814240-4664f9ee-9621-4e38-b38d-cdf60e14b25a.png">

I like how it's not tied to button clicks in the original diagram, but it is missing `Arrival` state and the transition from `FreeDrive` to `RoutePreview` is not solely based on the fact that `destination != null && route.isNotEmpty()`

abhishek1508 commented on libnavui-dropin/src/main/res/layout/mapbox_info_panel_active_guidance_layout.xml:
```diff
@@ -4,43 +4,51 @@
+                app:layout_constraintEnd_toEndOf="parent"
+                app:layout_constraintStart_toStartOf="parent" />
+
+            <include layout="@layout/mapbox_partial_trip_progress"
```
Let's remove `mapbox_partial_trip_progress` and `mapbox_partial_trip_progress_custom` and use `MapboxTripProgressView` here instead of `include`.

abhishek1508 commented on libnavui-dropin/src/main/res/layout/mapbox_info_panel_route_preview_layout.xml:
```diff
@@ -4,49 +4,31 @@
 
-        <TextView
-            android:id="@+id/routePreview"
+        <include layout="@layout/mapbox_partial_trip_progress"
```
Same as above.

abhishek1508 commented on libnavui-dropin/src/main/res/layout/mapbox_partial_trip_progress.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Let's remove this file and transfer the contents directly to the layouts using it

abhishek1508 commented on libnavui-dropin/src/main/res/layout/mapbox_partial_trip_progress_custom.xml:
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="utf-8"?>
```
This file can be deleted.

abhishek1508 commented on libnavui-dropin/src/main/res/values-night/colors.xml:
```diff
@@ -1,6 +1,13 @@
-    <color name="mapbox_button_fg_color">@color/colorOnSurface</color>
-    <color name="mapbox_button_bg_color">@color/colorSurface</color>
+    <color name="mapbox_drop_in_primary">@color/colorSecondary</color>
+    <color name="mapbox_drop_in_surface">#fff</color>
```
Let us not define any new colors here and use from `libnavui-resources/res/colors.xml`

abhishek1508 commented on libnavui-dropin/src/main/res/values-night/colors.xml:
```diff
@@ -1,6 +1,13 @@
-    <color name="mapbox_button_bg_color">@color/colorSurface</color>
+    <color name="mapbox_drop_in_primary">@color/colorSecondary</color>
+    <color name="mapbox_drop_in_surface">#fff</color>
+    <color name="mapbox_drop_in_onSurface">#3C3E3E</color>
```
Similar to above. Reason being theme is going to be controlled by NSS. Let's not introduce new color key_value pairs and rather keep using the ones we have so it's easy to just change them.

Same for all the other colors below

abhishek1508 commented on libnavui-dropin/src/main/res/values/colors.xml:
```diff
@@ -1,6 +1,14 @@
-    <color name="mapbox_button_fg_color">@color/colorOnSurface</color>
-    <color name="mapbox_button_bg_color">@color/colorSurface</color>
-    <color name="mapbox_button_border_color">#CDD0D0</color>
+    <color name="mapbox_drop_in_primary">@color/colorSecondary</color>
```
Same as above.

tomaszrybakiewicz commented on libnavui-dropin/src/main/res/layout/drop_in_navigation_view.xml:
```diff
@@ -48,34 +59,37 @@
+                tools:layout_height="200dp"
+                tools:background="#eee" />
 
-        <View
```
I removed this because view drop shadow should be done either by implementing [Outline](https://developer.android.com/reference/android/graphics/Outline) or by setting correct [elevation](https://material.io/design/environment/elevation.html#default-elevations). I've used 8dp elevation as suggested in Material Design guide.

tomaszrybakiewicz commented on libnavui-dropin/src/main/res/drawable/mapbox_map_pin_hole.xml:
```diff
@@ -0,0 +1,11 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
That's what's in Figma. Map pin has a "white" colored pin hole while Trip Progress pin has transparent one. This one is the latter one. I'll remove it since we are not implementing Figma Trip Progress designs.

tomaszrybakiewicz commented on libnavui-dropin/src/main/res/layout/drop_in_navigation_view.xml:
```diff
@@ -48,34 +59,37 @@
                 android:layout_margin="8dp"
                 app:layout_constraintTop_toBottomOf="@id/guidanceLayout"
                 app:layout_constraintBottom_toBottomOf="parent"
```
So you want this to be constrained only to TOP (maneuver view) and RIGHT (parent)? 
What about map orientation button in top right corner? Should we adjust actionListLayout top margin to ensure there is a space fo orientation button when it appears?
<img width="250" src="https://user-images.githubusercontent.com/2678039/156058189-0bdbb371-0aae-4602-a79c-691a08378098.png" />

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/extensions/coroutines/MapboxNavigation.kt:
```diff
@@ -0,0 +1,48 @@
+                        reasons,
+                        "Route request failed with: $reasons"
+                    )
+                    continuation.resumeWithException(error)
```
Why use a throwable to control execution flow instead of something like `Result`? The latter is explicit and won't allow for a mistake in the handling code.

tomaszrybakiewicz replied:
This extension method only converts callback API to a suspendible function. Not returning Result gives more flexibility over how exceptions are handled at the call site:

```kotlin
// use try catch
try {
  val v = navigation.requestRoutes(routeOptions)
} catch (e: RequestRoutesError) {
  // handle error here
}

// wrap it in a Result instance
val result: Result<RoutesWithOrigin> = runCatching {
    navigation.requestRoutes(routeOptions)
}
```

tomaszrybakiewicz replied:
@LukasPaczos I understand your concern. But, these extensions are not meant to be consumed by the general public.  These are internal, dropin module, low-level/data source APIs that are meant to be called by a domain layer [UseCases that already handle exceptions](https://github.com/mapbox/mapbox-navigation-android/blob/5b25af64d74897fe67e2886828dd5fe26f10414e/libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/usecase/UseCase.kt#L46).

cafesilencio replied:
There's no advantage to throwing exceptions. This and similar functions should return an Expected<Error, Type> like the majority of the API's in the SDK. The advantage is they are deterministic when exceptions aren't thrown.

tomaszrybakiewicz replied:
Since we are not going with coroutine based approach and want to stick to callbacks, I am removing this extension. This is no longer valid.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/coordinator/RouteManager.kt:
```diff
@@ -0,0 +1,30 @@
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+
+        viewModel.destination.observe { destination ->
+            val navState = viewModel.navigationState.value
+            if (enabledInStates.contains(navState) && destination != null) {
+                fetchAndSetRouteUseCase(destination.point)
+            }
+        }
```
This logic competes with the route request that happens inside of `InfoPanelFreeDriveComponent`. It'd be good to think about this redundancy and settle on an approach for route management.

tomaszrybakiewicz replied:
That's not a redundancy. 
The RouteManager fetches and sets new route on destination change only when in NavigationState.RoutePreview. 
`InfoPanelFreeDriveComponent` however will do it as a result of a direct user interaction.

LukasPaczos replied:
That's a surprising separation but I think nicely highlights the fact that we discussed yesterday - a need to clean this up and funnel the actions through one unit.

tomaszrybakiewicz replied:
Understood. How would you like me to change this? Either class or a sequence UML diagram would be very helpful here.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/coordinator/NavigationStateManager.kt:
````diff
@@ -0,0 +1,88 @@
+ *                                       ┗━━━━━━━━━━[ Arrival ]━━━━━━━━━━━━━━━━━┛
+ * ```
+ */
+internal class NavigationStateManager(
````
How do you envision developer being able to initialize the drop-in UI in any of the predefined states? It's not something that has to be implemented in this PR but we should think how to leave that door open.

tomaszrybakiewicz replied:
We didn't talk about how will host application interact with the Drop In UI, yet. 
I am still waiting for the "Technical document" (see my questions at the end of the "FAQ: Drop-In UI for Navigation SDK" doc).

Here are a couple of ways we can achieve this:
1. Expose single public method via DropInNavigationView
```kotlin
// free drive
navVIew.setInitialStateFreeDrive()
// free drive with POI
navVIew.setInitialStateFreeDrive(destination: Destination)

// route preview (we will fetch routes)
navView.setInitialStatePreview(destination: Destination)

// active guidance (we will fetch routes and start navigation session) 
navVIew.setInitialStateActiveGuidance(destination: Destination)
```

2.  Expose a mediator/controller/manager that can be used to retrieve current state and control nav UI.
```kotlin
navView.controller.setDestination(destination)
navView.controller.getNavigationState()
navView.controller.showRoutePreview(destination)
navView.controller.setVoiceInstructionsEnabled(enabled)
```

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/usecase/route/FetchRouteUseCase.kt:
```diff
@@ -0,0 +1,31 @@
+            .alternatives(true)
+            .build()
+
+        return navigation.requestRoutes(routeOptions).routes
```
This doesn't handle the exception that could be thrown in case of a request failure, refs https://github.com/mapbox/mapbox-navigation-android/pull/5506#discussion_r818536387.

tomaszrybakiewicz replied:
Superclass handles it.  [UseCase#invoke](https://github.com/mapbox/mapbox-navigation-android/blob/5b25af64d74897fe67e2886828dd5fe26f10414e/libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/usecase/UseCase.kt#L46)

tomaszrybakiewicz replied:
Since we are not going with UseCase approach, I am removing this class. This is no longer valid.

cafesilencio commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/usecase/FlowUseCase.kt:
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2018 Google LLC
```
I don't think we should be adding Google's source code to our SDK so casually. @LukasPaczos @zugaldia  does this need a more formal review?

zugaldia replied:
Apache is a permissive license which allows using code commercially as well as altering the code. Having said that, I'd be interested in learning more why we need to bring in this code here. I'll follow up with @tomaszrybakiewicz directly to understand. Let's not merge this PR until I have an opportunity to do that.

cafesilencio replied:
Don't we have to disclose the usage in our own license.md file? That gets difficult to track when it's source code in our SDK vs. a binary that we're importing.

tomaszrybakiewicz replied:
Since we are not using UseCase approach, I am removing these files. This is no longer valid.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/marker/LongPressMapComponent.kt:
```diff
@@ -0,0 +1,30 @@
+import com.mapbox.navigation.dropin.lifecycle.UIComponent
+import com.mapbox.navigation.dropin.model.Destination
+
+internal class LongPressMapComponent(
+    private val mapView: MapView,
+    private val viewModel: DropInNavigationViewModel
+) : UIComponent() {
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+        mapView.gestures.addOnMapLongClickListener(longClickListener)
+    }
+
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        super.onDetached(mapboxNavigation)
+        mapView.gestures.removeOnMapLongClickListener(longClickListener)
+    }
+
+    private val longClickListener = OnMapLongClickListener { point ->
+        viewModel.updateDestination(Destination(point))
+        false
+    }
+}
```
NICE i'm closing https://github.com/mapbox/mapbox-navigation-android/pull/5540

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/marker/LongPressMapComponent.kt:
```diff
@@ -0,0 +1,30 @@
+
+internal class LongPressMapComponent(
+    private val mapView: MapView,
+    private val viewModel: DropInNavigationViewModel
```
Can you remove the top level `DropInNavigationViewModel`

Would like to have the destination logic be part of a UIViewModel

tomaszrybakiewicz replied:
Which UIViewModel?

abhishek1508 replied:
It will depend on whether we want to store the destination across orientation changes. It personally makes sense to me store the destinations and waypoints. In that case we can create a `DestinationViewModel : UIViewModel<>` that would keep track of that data. Hence, when we create a `GeocodingComponent` that could depend on `DestinationViewModel`. If it seems too early to create this view model, we can store the destination data in `RouteViewModel`

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewContext.kt:
```diff
@@ -31,6 +48,63 @@ internal class DropInNavigationViewContext(
+
+    //region Use Cases
+
+    fun fetchRoutesUseCase() = FetchRouteUseCase(
```
Please convert theses logic cases to something that is a `UIViewModel`

tomaszrybakiewicz replied:
"Can you migrate this logic to one of the `UIViewModel` subclasses"?

Yes I can.

tomaszrybakiewicz commented:
@kmadsen @abhishek1508 I've rebased on top of main, replaced UseCase based implementation with UIViewModel one and removed unused classes. 
Please let me know if this is what you meant before I start writing unit tests.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewContext.kt:
```diff
@@ -31,6 +40,27 @@ internal class DropInNavigationViewContext(
         .withAboveLayerId(RouteLayerConstants.TOP_LEVEL_ROUTE_LINE_LAYER_ID)
         .build()
+
+    val onBackPressedCallback = object : OnBackPressedCallback(true) {
```
nit: Assuming we'll move this to BackPressManager in a future PR

tomaszrybakiewicz replied:
What does "nit" mean?

tomaszrybakiewicz replied:
I leave moving this up to you. I need it here to get back press working for InfoPanel.

kmadsen replied:
nitpick :). If you did it 💯 . back press could use some ownership, do you want to own it?

Issue with backpress is it requires configuration in the hosting activity. We'll want to fix that immediately

tomaszrybakiewicz replied:
Since you have a clear idea how you want this to be implemented, I think you should own it.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt:
```diff
@@ -21,25 +26,45 @@ internal class DropInNavigationViewModel : ViewModel() {
+    val _activeNavigationStarted = MutableStateFlow(false)
+    val activeNavigationStarted = _activeNavigationStarted.asStateFlow()
+
+    private val _onBackPressedEvent = MutableSharedFlow<Unit>(
```
Makes sense to throw this in a BackPressViewModel

tomaszrybakiewicz replied:
I leave BackPress handling to you. This is just a temp solution to get back press working.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/coordinator/NavigationStateManager.kt:
```diff
@@ -0,0 +1,105 @@
+    private val destinationFlow: Flow<Destination?>,
+    private val navigationState: MutableStateFlow<NavigationState>,
+    private val activeNavigationStarted: Flow<Boolean>
+) : UIComponent() {
```
We want this to be a `UIViewModel` as well

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt:
```diff
@@ -21,25 +26,45 @@ internal class DropInNavigationViewModel : ViewModel() {
 
+    @Suppress("PropertyName")
+    val _activeNavigationStarted = MutableStateFlow(false)
+    val activeNavigationStarted = _activeNavigationStarted.asStateFlow()
```
The _activeNavigationStarted and _navigationState make sense as part of the NavigationStateViewModel

tomaszrybakiewicz replied:
They already have been moved to RoutesState. I forgot to remove these.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt:
```diff
@@ -21,25 +26,45 @@ internal class DropInNavigationViewModel : ViewModel() {
-    fun updateState(state: NavigationState) {
-        _navigationState.value = state
+    fun onBackPressed() {
+        _onBackPressedEvent.tryEmit(Unit)
```
This should also be moved to the BackPressedViewModel

tomaszrybakiewicz replied:
I leave BackPress handling to you. This is just a temp solution to get back press working.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationUIBinders.kt:
```diff
@@ -9,4 +9,6 @@ class NavigationUIBinders(
     val maneuver: UIBinder = ManeuverViewBinder(),
     val roadName: UIBinder = RoadNameViewBinder(),
+    val infoPanelHeaderBinder: UIBinder? = null,
+    val infoPanelContentBinder: UIBinder? = null
```
+1

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/UIBinder.kt:
```diff
@@ -18,6 +18,8 @@ interface Binder<T> {
      */
     fun bind(value: T): MapboxNavigationObserver
+
+    fun unbind(value: T) = Unit
```
Curious on this one, why was this needed? The returned `MapboxNavigationObserver` calls `.onDetached`

tomaszrybakiewicz replied:
That's a fix for MapView being removed on detached.

kmadsen replied:
@tomaszrybakiewicz will you please revert the changes to `UIBinder` and `UICoordinator`

We will address the issue for removing the map separately. Doing this unbind impacts all UIBinder, but we have a special exception with the MapCoordinator.

tomaszrybakiewicz replied:
Done. I broke it again.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/util/CompositeMapboxNavigationObserver.kt:
```diff
@@ -0,0 +1,23 @@
+) : MapboxNavigationObserver {
+    private val observers = mutableListOf(*elements)
+
+    fun add(o: MapboxNavigationObserver) = observers.add(o)
+    fun remove(o: MapboxNavigationObserver) = observers.remove(o)
```
These should call `onAttached` and `onDetached`.

Do we use these functions?

tomaszrybakiewicz replied:
No.
Yes.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UICoordinator.kt:
```diff
@@ -21,33 +22,49 @@ import kotlinx.coroutines.launch
+                )
                 attachedObserver?.onDetached(mapboxNavigation)
-                attachedObserver = viewBinder.bind(viewGroup)
+                viewBinder?.also { unbind(it, viewGroup) }
```
Discussed offline, linking an issue this is resolving https://github.com/mapbox/mapbox-navigation-android/issues/5546

abhishek1508 replied:
It looks like trying to solve the issue with MapView not loading upon backgrounding/foregrounding is irrelevant to this ticket, if that is what we are trying to solve here. Can we rather identify these issue and cut tickets so that it could be handled separately thereby making the PR smaller. In other words if `unbind` is added to solve the [problem](https://github.com/mapbox/navigation-sdks/issues/1506) is it possible to not introduce it here and hande it separately?

tomaszrybakiewicz replied:
I fixed this issue as one of the first things when I started working on InfoPanel ticket. I am not going to rip this out just to re-add it again in a separate PR.
I'll refrain from fix anything that is outside active ticket scope in the future.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/util/CompositeMapboxNavigationObserver.kt:
```diff
@@ -0,0 +1,23 @@
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+internal class CompositeMapboxNavigationObserver(
```
This class can be easily abused with the add/remove functions. 

``` kotlin
val myComposite = CompositeMapboxNavigationObserver()
myComposite.add(otherObserver)
MapboxNavigatationApp.register(myComposite)
MapboxNavigatationApp.attach(lifecycleOwner)
myComposite.remove(otherObserver)
```

otherObserver.onDetached(..) will never be called. All of our observers are designed to have a guaranteed call to `onDetached`.

tomaszrybakiewicz replied:
That's not how composite listeners work. Anyway. I am removing this class.

tomaszrybakiewicz commented:
@kmadsen I think I addressed all your comments (besides BackPress ones). Could give it another look?

tomaszrybakiewicz commented:
@abhishek1508 @kmadsen Done. Anything else you would like me to change.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/routefetch/RoutesAction.kt:
```diff
@@ -0,0 +1,17 @@
+sealed class RoutesAction {
+    data class SetDestination(val destination: Destination?) : RoutesAction()
+    object FetchAndSetRoute : RoutesAction()
+    object StartNavigation : RoutesAction()
+    object StopNavigation : RoutesAction()
```
These should be part of `NavigationStateViewModel` actions?

tomaszrybakiewicz replied:
They depend on fetchRoute.

tomaszrybakiewicz replied:
^^ https://github.com/mapbox/mapbox-navigation-android/pull/5506/files/cb4796ff500924d993d3925d1f403d03547b38e2#r822093458

kmadsen commented on libnavui-dropin/src/main/res/layout/mapbox_info_panel_header_layout.xml:
```diff
@@ -0,0 +1,93 @@
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent" />
+
+        <com.mapbox.navigation.ui.tripprogress.view.MapboxTripProgressView
+            android:id="@+id/tripProgressView"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="20dp"
+            app:layout_constraintTop_toTopOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent" />
```
This can be addressed as tailwork

We want to allow the `MapboxTripProgressView` to be injected by itself. Right now we have the entire info_panel_header injected, but we want to be able to replace the `MapboxTripProgressView`.

So convert this into a FrameLayout and expose a `TripProgressBinder`

tomaszrybakiewicz replied:
Good idea. Would you like me to do this as a part of this PR?

kmadsen replied:
Yeah can you? That would be 💯

tomaszrybakiewicz commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/routefetch/RoutesViewModel.kt:
```diff
@@ -39,7 +54,23 @@ internal class RoutesViewModel : UIViewModel<Unit, RoutesAction>(Unit) {
+                    startNavigation(mapboxNavigation)
+                }
+                return state.copy(navigationStarted = true)
+            }
```
@kmadsen Any idea how can we better structure this?

kmadsen replied:
Yeah I was just staring at this too. I think this should be in a `NavigationStateViewModel`. It will have access to `MapboxNavigation` and routes. That view model will take over the `ReplayRoutesViewModel`, because sometimes we startReplayTripSessions

tomaszrybakiewicz replied:
OK, but how will `NavgationStateViewModel` fetch routes? 
When we start active navigation from Free Drive, routes might not be loaded for currently set Destination.
Should we add same method `fetchAndSetRoute` there as well?

kmadsen replied:
Wondering if we can resolve this with, making the destination be part of the `NavigationState` objects.

But I think it's fine and good to get this in and we can iterate on how to organize this logic

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/routefetch/RoutesViewModel.kt:
```diff
@@ -11,24 +12,38 @@ import com.mapbox.navigation.base.route.NavigationRouterCallback
 @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
-internal class RoutesViewModel : UIViewModel<Unit, RoutesAction>(Unit) {
+internal class RoutesViewModel(
+    private val navigationState: StateFlow<NavigationState>,
+    private val locationState: StateFlow<Location?>
```
We have a goal to be able to expose our view model implementations publically.

Ideally we delete these 2 lines. It's better to depend on the view models (e.g., NavigationStateViewModel and LocationViewModel)

But we also, want to minimize the dependencies. So best case, we are able to remove all the dependencies from this view model.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/routefetch/RoutesViewModel.kt:
```diff
@@ -49,7 +80,53 @@ internal class RoutesViewModel : UIViewModel<Unit, RoutesAction>(Unit) {
+        destination: Destination?,
+        cb: (() -> Unit)? = null
+    ) {
+        val from = locationState.value?.toPoint()
```
This looks like a valid case to depend on `LocationViewModel` for now. Replay the locationState flow with the state flow in `LocationViewModel`

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/navigation/NavigationStateViewModel.kt:
```diff
@@ -0,0 +1,119 @@
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+
+        mainJobControl.scope.launch {
+            routesState.collect {
+                val state = getNavigationState(
+                    it.destination,
+                    it.navigationStarted,
+                    mapboxNavigation.getRoutes()
+                )
+                updateState(state)
+            }
+        }
+
+        mainJobControl.scope.launch {
+            mapboxNavigation.flowRoutesUpdated().collect {
+                val state = getNavigationState(
+                    routesState.value.destination,
+                    routesState.value.navigationStarted,
+                    it.routes
+                )
+                updateState(state)
+            }
+        }
+
+        mainJobControl.scope.launch {
+            mapboxNavigation.flowOnFinalDestinationArrival().collect {
+                val inActiveNav =
+                    state.value == NavigationState.ActiveNavigation
+                val arrived = it.currentState == RouteProgressState.COMPLETE
+                if (inActiveNav && arrived) {
+                    updateState(NavigationState.Arrival)
+                }
+            }
+        }
```
Convert each of these into `Action`s with parameters. You will be able to remove the circular dependency. Removing the dependency on the ViewModels

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/navigation/NavigationStateViewModel.kt:
```diff
@@ -0,0 +1,119 @@
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+
+        mainJobControl.scope.launch {
+            routesState.collect {
+                val state = getNavigationState(
+                    it.destination,
+                    it.navigationStarted,
+                    mapboxNavigation.getRoutes()
+                )
+                updateState(state)
+            }
+        }
+
+        mainJobControl.scope.launch {
+            mapboxNavigation.flowRoutesUpdated().collect {
+                val state = getNavigationState(
+                    routesState.value.destination,
+                    routesState.value.navigationStarted,
+                    it.routes
+                )
+                updateState(state)
+            }
+        }
+
+        mainJobControl.scope.launch {
+            mapboxNavigation.flowOnFinalDestinationArrival().collect {
+                val inActiveNav =
+                    state.value == NavigationState.ActiveNavigation
+                val arrived = it.currentState == RouteProgressState.COMPLETE
+                if (inActiveNav && arrived) {
+                    updateState(NavigationState.Arrival)
+                }
+            }
+        }
```
We need to establish a definition for what the mapping is for each `NavigationState`

It will allow us to delete the dependencies for `NavigationStateViewModel` and the circular dependency created with `RoutesViewModel`

The definition for "why exactly navigation state" needs RoutesViewModel will help us decide how to organize it.

After digging, the reason we cannot reduce the NavigationState without `RoutesViewModel` is because we have a RoutePreview state. The RoutePreview state is not supported inside `MapboxNavigation` because, the moment you set a route, you start ActiveNavigation.

That's the root issue, solution appears ambiguous on how exactly we handle it in drop-in. But we do have "preferable" solutions. And we prefer to remove the circular dependency

kmadsen replied:
Here is an idea for removing the circular dependency.

Remove all the logic from the NavigationStateViewModel. Add a NavigationStateComponent which depends on the RoutesViewModel and the NavigationStateViewModel. It then decides how to update the state 

https://github.com/mapbox/mapbox-navigation-android/compare/tr-%231438-dropin-info-panel...km-info-panel-refactor?expand=1

tomaszrybakiewicz replied:
I already had this with a different name (NavigationStateManager) and references to each individual Flow instead of ViewModels.
https://github.com/mapbox/mapbox-navigation-android/pull/5506/commits/81d4ab1d6998b724c34f9609813f7d30a0eb870a#diff-bb50b0e1f2fbbb2d2ccf984bceb5baaa41496d5f6ac1818924c68d9ec7c09e93

I'll merge your branch to mine and take it from there.

tomaszrybakiewicz commented:
@kmadsen I've merged your changes and added support for custom TripProgress binder. Let me know if there is anything else you would like to change.

tomaszrybakiewicz commented:
@kmadsen @abhishek1508 I've rebased and squashed on `main`, and added unit tests. Let me know if there is anything else I should change.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UIViewModel.kt:
```diff
@@ -21,10 +21,15 @@ import kotlinx.coroutines.launch
 @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
-abstract class UIViewModel<State, Action>(initialState: State) : MapboxNavigationObserver {
+abstract class UIViewModel<State, Action>(
+    mutableState: MutableStateFlow<State>
```
@tomaszrybakiewicz is this a residual change when resolving the circular dependency?

We don't have a reason to do this yet so this should be reverted

kmadsen replied:
Let's merge though, I can follow up on this

kmadsen reviewed (approved):
Thanks for addressing feedback! 

A lot of good stuff in this pull request!

 - Added marker for destination
 - Added customizable trip progress view
 - Added customizable info panel content
 - Added info panel that matches designs
 - Added navigation state logic
 - Added navigation state to back stack

🪂

kmadsen commented:
Looks like it's failing robo-tests though 🤔 

https://console.firebase.google.com/u/0/project/mapbox-navigation-android/crashlytics/app/android:com.mapbox.navigation.examples/issues/d5902400f02aada360ca13dc03a8b8a6?time=last-seven-days&sessionEventKey=62292805003800015602928A934E41B0_1651674816136337918

```
Fatal Exception: java.lang.NullPointerException: viewGroup.findViewById(R.id.infoPanelHeader) must not be null
       at com.mapbox.navigation.dropin.binder.infopanel.InfoPanelBinder.bind(InfoPanelBinder.java:18)
       at com.mapbox.navigation.dropin.binder.infopanel.InfoPanelBinder.bind(InfoPanelBinder.java:2)
       at com.mapbox.navigation.dropin.lifecycle.UICoordinator$onAttached$1$invokeSuspend$$inlined$collect$1.emit(UICoordinator.java:24)
       at com.mapbox.navigation.dropin.coordinator.InfoPanelCoordinator$flowViewBinders$$inlined$map$1$2.emit(InfoPanelCoordinator.java:84)
       at kotlinx.coroutines.flow.StateFlowImpl.collect(StateFlowImpl.java:201)
       at com.mapbox.navigation.dropin.coordinator.InfoPanelCoordinator$flowViewBinders$$inlined$map$1.collect(InfoPanelCoordinator.java:9)
       at com.mapbox.navigation.dropin.lifecycle.UICoordinator$onAttached$1.invokeSuspend(UICoordinator.java:47)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(BaseContinuationImpl.java:9)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.java:129)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:223)
       at android.app.ActivityThread.main(ActivityThread.java:7664)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

kmadsen commented:
I rebased this, not sure what was causing ui-robo-tests to fail. I also tried to build a release version of the app in case it was some kind of `minifyEnabled` issue. But no luck, am not able to make it fail locally yet.

Maybe a rebase re-run will fix it

kmadsen commented:
Adding a video for initial launch. It takes 5 seconds to come from the ocean. Would be nice to make the start up faster. But also happy to discuss and make it follow up because this pull request is also fixing a lot

https://user-images.githubusercontent.com/3021882/157573016-7fe64948-3588-49f3-90a9-36de3468802e.mp4

kmadsen commented:
Another issue I'm seeing is after setting up a route the current location is not visible. After we add a destination marker, I think this view will expect to see both the current location and the destination marker

https://user-images.githubusercontent.com/3021882/157573794-b793dd7c-8881-49b2-9a3f-965d387bd0fe.mp4

tomaszrybakiewicz commented:
@kmadsen I don't have access to the Firebase Console.
How would you like to handle this? Should we just merge and deal with this later?

kmadsen commented:
Well it's green now. The test that I'm seeing that is having issues is a LANDSCAPE. Considering that unrelated to this PR. The issue was not reproducible anyways so there's no action item here. Let's merge party 🚢 

![Screen Shot 2022-03-10 at 8 40 57 AM](https://user-images.githubusercontent.com/3021882/157711675-4c57642f-592d-400c-8baf-d443c962c3f3.png)

![Screen Shot 2022-03-10 at 8 46 22 AM](https://user-images.githubusercontent.com/3021882/157712701-f86b0b02-3dad-4aee-b6cc-bbbfea8ad20c.png)

kmadsen commented:
OH I see what is happening now. This actually fixed it by accident https://github.com/mapbox/mapbox-navigation-android/pull/5553

After we rebased. Before the Binder would remove the two views when the app calls onStop.
```
infoPanelHeader
infoPanelContent
```

kmadsen commented:
The fix we would have had, would be to add these to a binder to ensure they are re-inflated when the `bind` is called.
```
infoPanelHeader
infoPanelContent
```

That will ensure this can still work because the view will be re-inflated.
